utils/global-state-update-gen/src/generic.rs

```diff
@@ -0,0 +1,284 @@
+    let update = get_update(reader, config);
+    for (key, value) in update {
+        print_entry(&key, &value);
+    }
+}
+
+fn process_transfers<T: StateReader>(state: &mut StateTracker<T>, transfers: &[Transfer]) {
+    for transfer in transfers {
+        state.execute_transfer(transfer);
+    }
+}
+
+fn update_account_balances<T: StateReader>(
+    state: &mut StateTracker<T>,
+    accounts: &[AccountConfig],
+) {
+    for account in accounts {
+        let target_balance = if let Some(balance) = account.balance {
+            balance
+        } else {
+            continue;
+        };
+        let account_hash = account.public_key.to_account_hash();
+        if let Some(account) = state.get_account(&account_hash) {
+            state.set_purse_balance(account.main_purse(), target_balance);
+        } else {
+            state.create_account(account_hash, target_balance);
+        }
+    }
+}
+
+fn update_auction_state<T: StateReader>(
+    state: &mut StateTracker<T>,
+    accounts: &[AccountConfig],
+    only_listed_validators: bool,
+) {
+    // Read the old SeigniorageRecipientsSnapshot
+    let (validators_key, old_snapshot) = state.read_snapshot();
+
+    // Create a new snapshot based on the old one and the supplied validators.
+    let new_snapshot = if only_listed_validators {
+        gen_snapshot_only_listed(
+            *old_snapshot.keys().next().unwrap(),
+            old_snapshot.len() as u64,
+            accounts,
+        )
+    } else {
+        gen_snapshot_from_old(old_snapshot.clone(), accounts)
+    };
+
+    if new_snapshot != old_snapshot {
+        // Save the write to the snapshot key.
+        state.write_entry(
+            validators_key,
+            StoredValue::from(CLValue::from_t(new_snapshot.clone()).unwrap()),
+        );
+
+        let validators_diff = validators_diff(&old_snapshot, &new_snapshot);
+
+        add_and_remove_bids(
+            state,
+            &validators_diff,
+            &new_snapshot,
+            only_listed_validators,
+        );
+
+        state.remove_withdraws(&validators_diff.removed);
+    }
+}
+
+/// Generates a new `SeigniorageRecipientsSnapshot` based on:
+/// - The starting era ID (the era ID at which the snapshot should start).
+/// - Count - the number of eras to be included in the snapshot.
+/// - The list of configured accounts.
+fn gen_snapshot_only_listed(
+    starting_era_id: EraId,
+    count: u64,
+    accounts: &[AccountConfig],
+) -> SeigniorageRecipientsSnapshot {
+    let mut new_snapshot = BTreeMap::new();
+    let mut era_validators = BTreeMap::new();
+    for account in accounts {
+        // don't add validatord with zero stake to the snapshot
+        let stake = match account.stake {
+            Some(stake) if stake != U512::zero() => stake,
+            _ => continue,
+        };
+        let seigniorage_recipient =
+            SeigniorageRecipient::new(stake, Default::default(), Default::default());
```

Comment:
What happens to the delegators?

I imagine that we'd want to preserve the good validators' entries, including all their delegators?

Reply:
Good point. I was kind of ignoring the delegators, but I probably shouldn't. I'll fix it :+1: